pinot-core/src/main/java/org/apache/pinot/core/data/IndexedTableImpl.java

```diff
@@ -0,0 +1,169 @@
+import org.apache.pinot.common.utils.DataSchema;
+import org.apache.pinot.core.query.aggregation.function.AggregationFunction;
+import org.apache.pinot.core.query.aggregation.function.AggregationFunctionUtils;
+
+
+/**
+ * {@link IndexedTable} implementation for aggregating TableRecords based on combination of keys
+ */
+public class IndexedTableImpl implements IndexedTable {
+
+  /** Factor used to add buffer to maxCapacity of the Collection used **/
+  private static final double BUFFER_FACTOR = 1.2;
+  /** Factor used to decide eviction threshold **/
+  private static final double EVICTION_FACTOR = 1.1;
+
+  private List<TableRecord> _records = new ArrayList<>();
+  private ConcurrentMap<TableRecord, Integer> _lookupTable;
+  private ReentrantReadWriteLock _readWriteLock;
+
+  private DataSchema _dataSchema;
+  private List<AggregationInfo> _aggregationInfos;
+  private List<AggregationFunction> _aggregationFunctions;
+  private List<SelectionSort> _orderBy;
+  private int _evictCapacity;
+  private int _bufferedCapacity;
+
+  @Override
+  public void init(@Nonnull DataSchema dataSchema, List<AggregationInfo> aggregationInfos, List<SelectionSort> orderBy,
+      int maxCapacity) {
+    _dataSchema = dataSchema;
+    _aggregationInfos = aggregationInfos;
+    _orderBy = orderBy;
+
+    // maintain a buffer size 20% more than max capacity
+    // this is to ensure bottom records get a fair chance
+    _bufferedCapacity = (int) (maxCapacity * BUFFER_FACTOR);
+    // after eviction, bring down size to 10% more than max capacity
+    _evictCapacity = (int) (maxCapacity * EVICTION_FACTOR);
+
+    _lookupTable = new ConcurrentHashMap<>(_bufferedCapacity);
+    _readWriteLock = new ReentrantReadWriteLock();
+
+    if (CollectionUtils.isNotEmpty(aggregationInfos)) {
+      _aggregationFunctions = new ArrayList<>(aggregationInfos.size());
+      for (AggregationInfo aggregationInfo : aggregationInfos) {
+        _aggregationFunctions.add(
+            AggregationFunctionUtils.getAggregationFunctionContext(aggregationInfo).getAggregationFunction());
+      }
+    }
+  }
+
+  @Override
+  public boolean upsert(@Nonnull TableRecord newRecord) {
+
+    Object[] keys = newRecord._keys;
+    Preconditions.checkNotNull(keys, "Cannot upsert record with null keys");
+
+    if (size() >= _bufferedCapacity && !_lookupTable.containsKey(newRecord)) {
+      _readWriteLock.writeLock().lock();
+      try {
+        if (size() >= _bufferedCapacity) {
+          sort();
+          _records = _records.subList(0, _evictCapacity);
+          rebuildLookupTable();
+        }
+      } finally {
+        _readWriteLock.writeLock().unlock();
+      }
+    }
+
+    _readWriteLock.readLock().lock();
+    try {
+      _lookupTable.compute(newRecord, (k, index) -> {
+        if (index == null) {
+          index = size();
+          _records.add(newRecord);
+        } else {
+          if (CollectionUtils.isNotEmpty(_aggregationFunctions)) {
```

Comment:
is this needed?

Reply:
not for group by, but if and when this is used for distinct, it'll be needed. Do you prefer I remove it for now?

Reply:
yes. will be good to keep this code simple with least branching/if checks.